File: Linklist/reverse_in_group.rb
# Algorithm
#
# 1. take 3 vars i.e. curr, prev, next
# 2. On every recursion call, reverse the group iteratively and return prev (new head)
Node = Struct.new(:key, :next)

def pp(node)
  curr = node
  res = []
  while curr != nil
    res << curr.key
    curr = curr.next
  end
  puts res.join(" -> ")
end

def reverse_in_group(head, k)
  prev = nil
  curr = head

  i = 0

  while curr && i < k
    _next = curr.next
    curr.next = prev
    prev = curr
    curr = _next
    i += 1
  end


  head.next = reverse_in_group(curr, k) if curr


  prev

end


[1,2,3,4].each do |k|

  head = Node.new(1)
  head.next = Node.new 2
  head.next.next = Node.new 3
  head.next.next.next = Node.new 4
  head.next.next.next.next = Node.new 5
  head.next.next.next.next.next = Node.new 6
  head.next.next.next.next.next.next = Node.new 7
  head.next.next.next.next.next.next.next = Node.new 8


  puts "Original: "
  pp head
  head = reverse_in_group head, k
  puts "Reversed in group of #{k}"
  pp head
  puts
end

File: Linklist/ReverseInGroup.java
class ReverseInGroup {

	public static void main(String[] args) {
		Node head = new Node(1);
		head.next = new Node(2);
		head.next.next = new Node(3);
		head.next.next.next = new Node(4);
		head.next.next.next.next = new Node(5);
		head.next.next.next.next.next = new Node(6);
		head.next.next.next.next.next.next = new Node(7);
		head.next.next.next.next.next.next.next = new Node(8);
		System.out.print("List : ");
		printList(head);

		int k = 3;
		head = reverse(head, k);
		System.out.print("\nReversed by ("+ k +") : ");
		printList(head);

	}

    public static Node reverse(Node node, int k) {
        if(node == null || node.next == null)
            return node;

        /*
            1 2 3 4 5 6 7 8
                ^
        */
        Node curr = node;
        for(int i = 1; i < k && curr != null; i++) {
            curr = curr.next;
        }

        Node next = null;
        if(curr != null) {
            next = curr.next;
            curr.next = null;
        }

        Node head = reverse(node);
        node.next = reverse(next, k);

        return head;

    }

    static Node reverse(Node node) {
        if(node == null || node.next == null)
            return node;

        Node head = reverse(node.next);
        node.next.next = node;
        node.next = null;

        return head;
    }

    static void printList(Node head) {
    	while(head != null) {
    		System.out.print(head.data + " ");
    		head = head.next;
    	}
    	System.out.println();
    }
}

class Node {
    int data;
    Node next;
    Node(int key) { data = key; }
}